Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,221 @@
+//
+//  VersionFile.swift
+//  Carthage
+//
+//  Created by Jason Boyle on 8/11/16.
+//  Copyright © 2016 Carthage. All rights reserved.
+//
+
+import Foundation
+import Argo
+import Curry
+
+private struct CachedFramework {
+	let name: String
+	let sha1: String
+	
+	static let nameKey = "name"
+	static let sha1Key = "sha1"
+	
+	func toJSONObject() -> AnyObject {
+		return [
+			CachedFramework.nameKey: name,
+			CachedFramework.sha1Key: sha1
+		]
+	}
+}
+
+extension CachedFramework: Decodable {
+	static func decode(j: JSON) -> Decoded<CachedFramework> {
+		return curry(self.init)
+			<^> j <| CachedFramework.nameKey
+			<*> j <| CachedFramework.sha1Key
+	}
+}
+
+private struct CachedPlatform {
+	let commitish: String
+	let cachedFrameworks: [CachedFramework]
+	
+	static let commitishKey = "commitish"
+	static let cachedFrameworksKey = "cachedFrameworks"
+	
+	func toJSONObject() -> AnyObject {
+		let array = cachedFrameworks.map { $0.toJSONObject() }
+		return [
+			CachedPlatform.commitishKey: commitish,
+			CachedPlatform.cachedFrameworksKey: array
+		]
+	}
+}
+
+extension CachedPlatform: Decodable {
+	static func decode(j: JSON) -> Decoded<CachedPlatform> {
+		return curry(self.init)
+			<^> j <| CachedPlatform.commitishKey
+			<*> j <|| CachedPlatform.cachedFrameworksKey
+	}
+}
+
+private struct VersionFile {
+	let macOS: CachedPlatform?
+	let iOS: CachedPlatform?
+	let watchOS: CachedPlatform?
+	let tvOS: CachedPlatform?
+	
+	func cacheForPlatform(platform: Platform) -> CachedPlatform? {
+		switch platform {
+		case .Mac:
+			return macOS
+		case .iOS:
+			return iOS
+		case .watchOS:
+			return watchOS
+		case .tvOS:
+			return tvOS
+		}
+	}
+	
+	func cachedPlatforms() -> Set<Platform> {
+		return Set(Platform.supportedPlatforms
+			.flatMap { platform in
+				return self.cacheForPlatform(platform) != nil ? [platform] : []
+			})
+	}
+	
+	func toJSONObject() -> AnyObject {
+		return Platform.supportedPlatforms.reduce([String: AnyObject](), combine: { (dict, platform) in
+			var dict = dict
+			dict[platform.rawValue] = cacheForPlatform(platform)?.toJSONObject()
+			return dict
+		})
+	}
+}
+
+extension VersionFile: Decodable {
+	static func decode(j: JSON) -> Decoded<VersionFile> {
+		return curry(self.init)
+			<^> j <|? Platform.Mac.rawValue
+			<*> j <|? Platform.iOS.rawValue
+			<*> j <|? Platform.watchOS.rawValue
+			<*> j <|? Platform.tvOS.rawValue
+	}
+}
+
+private typealias VersionFileData = [String: [String: AnyObject]]
+
+/// Creates a version file for the current dependency in the
+/// Carthage/Build directory which associates its commitish with
+/// the SHA1s of the built frameworks for each platform in order
+/// to allow those frameworks to be skipped in future builds.
+///
+/// Returns true if the version file was successfully created.
+public func createVersionFileForDependency(dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, buildProductURLs: [NSURL], rootDirectoryURL: NSURL) -> Bool {
```

Comment:
1. This seems like it should, at the very least, use a `write(to url: URL)` method on `VersionFile`.

2. It should return a `SignalProducer` so that it can be a non-blocking call.